File: javascript/lesson 10/howSumRecursive.js
function howSum(targetSum, numbers) {
    return howSumImpl(targetSum, numbers, new Array(targetSum + 1));
}

function howSumImpl(targetSum, numbers, dp) {
    if (targetSum < 0) return null;
    if (targetSum == 0) return [];
    if (dp[targetSum]) return dp[targetSum];        
    
    for (let number of numbers) {
        const current = howSumImpl(targetSum - number, numbers, dp);
        if (current) {
            current.push(number);
            dp[targetSum] = current;
            return dp[targetSum];
        }
    }
    
    return null;
}

File: python/lesson 10/how_sum_recursive.py
from typing import List

def how_sum(target_sum: int, numbers: List[int]) -> List[int]:
    return how_sum_impl(target_sum, numbers, [None] * (target_sum + 1))

def how_sum_impl(target_sum: int, numbers: List[int], dp: List[List[int]]) -> List[int]:
    if target_sum < 0: return None
    if target_sum == 0: return [] 
    if dp[target_sum] != None: return dp[target_sum]
    
    for number in numbers:
        current = how_sum_impl(target_sum - number, numbers, dp)
        if current != None:
            current.append(number)
            dp[target_sum] = current
            return dp[target_sum]
        
    return None
